### PR TITLE
Review dependantRoot usage

### PR DESCRIPTION
**Pull request by dapplion**
**Motivation**

General clean-up, drop the need to have an intermediary event and re-use code from findAttesterDependentRoot

**Description**

- Generalize `findAttesterDependentRoot` to support previous epoch and current epoch
- Use getDependantRoot to compute the head event dependant roots
- Cache shuffling dependant root and re-use in API
- Make dependantRoot hex in API to prevent unnecessary type conversions

---

@tuyennhv testing this branch with 

```
lodestar.js dev --startValidators 0..7 --api.rest.enabled --api.rest.api '*'
```

I get a lot of logs for re-orgs so some logic must be wrong somewhere between the validator consumer or beacon node producer. For every slot after epoch > 0
```
Eph 1/0 0.061[]                 warn: Potential next epoch attester duties reorg slot=8, dutyEpoch=2, priorDependentRoot=0xdca272c39cb493f10e3c400af6267632a78172f6636894d0e6166d25d8cca104, newDependentRoot=0x8649261fe405bcf7672e12d1b1a3e29aae3be28d228204d108af49c42bc6d87f
```
and for every slot after epoch > 1

```Eph 2/0 0.042[]                 warn: Potential next epoch attester duties reorg slot=16, dutyEpoch=3, priorDependentRoot=0x8649261fe405bcf7672e12d1b1a3e29aae3be28d228204d108af49c42bc6d87f, newDependentRoot=0x41fc2d22696e18c614a33dcc5d1a44d335a95fa5b48a2682c2c391678fe43a29
Eph 2/0 0.046[]                 warn: Potential current epoch attester duties reorg slot=16, dutyEpoch=2, priorDependentRoot=0xdca272c39cb493f10e3c400af6267632a78172f6636894d0e6166d25d8cca104, newDependentRoot=0x8649261fe405bcf7672e12d1b1a3e29aae3be28d228204d108af49c42bc6d87f
```

would you mind taking a look? I don't want to create a regression for the issue at https://github.com/ChainSafe/lodestar/issues/3211

### CODE REVIEW

File: packages/state-transition/src/util/epochShuffling.ts [hunk 1]
```diff
@@ -99,6 +99,6 @@ export function computeEpochShuffling(
     shuffling,
     committees,
     committeesPerSlot,
-    dependantRoot: toHex(attesterShufflingDecisionRoot(state, epoch - 1)),
```
Comment (twoeths) [inline packages/state-transition/src/util/epochShuffling.ts]: @dapplion this works fine in my test, please try again at your side

Comment (dapplion) [reply packages/state-transition/src/util/epochShuffling.ts]: Thank you! Don't see the `warn: Potential **** epoch attester duties reorg` in either epoch 0,1,2+

File: packages/fork-choice/src/forkChoice/forkChoice.ts [hunk 2]
```diff
@@ -754,40 +754,56 @@ export class ForkChoice implements IForkChoice {
...
-    throw Error(
-      `Not able to find attester dependent root for head ${headBlockHash}, slot ${slot}, pivotSlot ${pivotSlot}, last tried slot ${block.slot}`
-    );
+  getDependantRoot(block: ProtoBlock, epochDifference: EpochDifference): RootHex {
+    // The navigation at the end of the while loop will always progress backwards,
+    // jumping to a block with a strictly less slot number. So the condition `blockEpoch < atEpoch`
+    // is guaranteed to happen. Given the use of target blocks for faster navigation, it will take
+    // at most `2 * (blockEpoch - atEpoch + 1)` iterations to find the dependant root.
+
+    const beforeSlot = block.slot - (block.slot % SLOTS_PER_EPOCH) - epochDifference * SLOTS_PER_EPOCH;
```
Comment (nazarhussain) [inline packages/fork-choice/src/forkChoice/forkChoice.ts:777]: Would be better to improve the readability of code by using helper methods. I am from the philosophical group believing that **comments are the code smell**. 

```ts
const beforeSlot = computeEndSlotAtEpoch(block.slot - epochDifference);
```

File: packages/fork-choice/src/forkChoice/forkChoice.ts [hunk 3]
```diff
@@ -754,40 +754,56 @@ export class ForkChoice implements IForkChoice {
...
+
+    // Special case close to genesis block, return the genesis block root
+    if (beforeSlot <= 0) {
+      const genesisBlock = this.protoArray.nodes[0];
```
Comment (nazarhussain) [inline packages/fork-choice/src/forkChoice/forkChoice.ts:781]: If that's the case may be we make the `getNodeFromIndex` function `public` and then use it here.

File: packages/fork-choice/src/forkChoice/forkChoice.ts [hunk 4]
```diff
@@ -754,40 +754,56 @@ export class ForkChoice implements IForkChoice {
...
+      if (genesisBlock === undefined || genesisBlock.slot !== 0) {
+        throw Error("Genesis block not available");
+      }
+      return genesisBlock.blockRoot;
+    }
+
+    // eslint-disable-next-line no-constant-condition
+    while (true) {
```
Comment (nazarhussain) [inline packages/fork-choice/src/forkChoice/forkChoice.ts:789]: The earlier logic was explicit of not looking too far in the history. Should be limiting it to `MAX_SEED_LOOKAHEAD` make sense here?

**Review by nazarhussain — CHANGES_REQUESTED** [review]

**Comment by dapplion** [comment]
Closing for separate PRs
- https://github.com/ChainSafe/lodestar/pull/4556
- https://github.com/ChainSafe/lodestar/pull/4555
- https://github.com/ChainSafe/lodestar/pull/4554
- https://github.com/ChainSafe/lodestar/pull/4553

File: packages/fork-choice/src/forkChoice/forkChoice.ts (hunk 4 above)
Comment (dapplion) [reply packages/fork-choice/src/forkChoice/forkChoice.ts:789]: `MAX_SEED_LOOKAHEAD` is a concept foreign to the fork-choice at this level. This function is generic with any EpochDifference, which is an enum with 0 or -1 just convenience

File: packages/fork-choice/src/forkChoice/forkChoice.ts (hunk 2 above)
Comment (dapplion) [reply packages/fork-choice/src/forkChoice/forkChoice.ts:777]: Agree in general, each tool (including comments) must be used where appropiate. Your code snippet is wrong, mixing slot and epoch. The right implementation is:
```ts
const beforeSlot = computeEndSlotAtEpoch(computeEpochAtSlot(block.slot) - epochDifference))
```
which isn't that much readable TBH

**Pull request closed without merging**